src/levanter/schedule.py

```diff
@@ -0,0 +1,122 @@
+import typing
+from dataclasses import dataclass
+from typing import Sequence, TypeVar
+
+
+T = TypeVar("T")
+
+
+@dataclass
+class ScheduleStep(typing.Generic[T]):
+    until: int
+    value: T
+
+
+IntSchedule = list[ScheduleStep[int]]
+
+BIG_INT = 2**63 - 1
+
+
+def value_at_step(schedule_or_t: Sequence[ScheduleStep[T]] | T, step: int) -> T:
+    """
+    Given a schedule or a single value, return the value at the given step.
+
+    """
+    if not isinstance(schedule_or_t, Sequence) or (schedule_or_t and not isinstance(schedule_or_t[0], ScheduleStep)):
+        return schedule_or_t  # type: ignore
+
+    for i, step_ in enumerate(schedule_or_t):
+        if step < step_.until or step_.until == -1:
+            return step_.value
+
+    return schedule_or_t[-1].value
+
+
+def validate_schedule_sorted(schedule: Sequence[ScheduleStep[T]]):
+    for i in range(1, len(schedule)):
+        # the last step can be -1 to indicate that the value should be held indefinitely
+        if schedule[i].until < schedule[i - 1].until:
+            if i == len(schedule) - 1 and schedule[i].until == -1:
+                continue
+            raise ValueError(f"Schedule is not sorted at index {i}")
+
+
+@dataclass
+class BatchSegment:
+    start: int  # The training step at which this batch size starts.
+    until: int | float  # The training step at which this batch size stops. If -1 or inf, the segment is open-ended.
+    value: int  # The batch size for steps before 'until'.
+    offset: int  # The cumulative number of data points processed up to the start of this segment.
+
+
+class BatchSchedule:
+    """
+    A class that schedules the batch size for each training step. The schedule can be either a fixed batch size or a
+    schedule of batch sizes over time. The schedule is defined by a list of steps, each of which has a batch size and
+    a step at which that batch size starts.
+
+    This class also tracks cumulative data offsets and can return the "local" indices for a step and a sharding
+    """
+
+    def __init__(self, schedule: int | IntSchedule):
+        self.schedule = schedule
+
+        # Precompute the segments of the schedule.
+        if isinstance(schedule, int):
+            self.segments = [BatchSegment(0, BIG_INT, schedule, 0)]
```

Comment:
Why BIG_INT and not Inf here? Given the lower level class lists support for inf, seems odd (while unlikely to ever come up) to use this internally.

Reply:
was switching to big_int